src/dlfcn.c

```diff
@@ -467,7 +472,170 @@ char *dlerror( void )
     return error_buffer;
 }
 
-#ifdef SHARED
+/* taken from http://bandido.ch/programming/Import_Address_Table_Hooking.pdf */
+static IMAGE_IMPORT_DESCRIPTOR* get_import_address_table( HMODULE module, DWORD *size )
+{
+    IMAGE_DOS_HEADER* dosHeader = (IMAGE_DOS_HEADER*)module;
+    if ( dosHeader->e_magic != 0x5A4D )
+        return NULL;
+    IMAGE_OPTIONAL_HEADER* optionalHeader = (IMAGE_OPTIONAL_HEADER*)
+        ((BYTE*)module + dosHeader->e_lfanew + 24);
+    if ( optionalHeader->Magic != IMAGE_NT_OPTIONAL_HDR_MAGIC )
+        return NULL;
+    if ( optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size == 0 ||
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress == 0 )
+        return NULL;
+    *size = optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size;
+    return (IMAGE_IMPORT_DESCRIPTOR*)((BYTE*)module +
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress);
+}
+
+/*
+ * return symbol name for a given address
+ */
+static const char *get_symbol_name( HMODULE module, IMAGE_IMPORT_DESCRIPTOR *iid, void *addr )
+{
+    int i;
+    for(i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++) {
+        BYTE *base = (BYTE *)module; /* required to have correct calculations */
+        IMAGE_THUNK_DATA *thunkILT = (IMAGE_THUNK_DATA*)(iid[i].Characteristics + base);
+        IMAGE_THUNK_DATA *thunkIAT = (IMAGE_THUNK_DATA*)(iid[i].FirstThunk + base);
+        for(; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++) {
+            if (IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal))
+              continue;
+            if ((void *)thunkIAT->u1.Function != addr)
+              continue;
+            IMAGE_IMPORT_BY_NAME *nameData = (IMAGE_IMPORT_BY_NAME*)(thunkILT->u1.AddressOfData + base);
+            return (const char *)nameData->Name;
+        }
+   }
+   return NULL;
+}
+
+static BOOL is_valid_address( void *addr )
+{
+    if( addr == NULL )
+        return FALSE;
+
+    /* check valid pointer */
+    MEMORY_BASIC_INFORMATION info;
+    SIZE_T result = VirtualQuery( addr, &info, sizeof( info ) );
+    if( result == 0 || info.AllocationBase == NULL || info.AllocationProtect == 0 || info.AllocationProtect == PAGE_NOACCESS )
+        return FALSE;
+    return TRUE;
+}
+
+/*
+ * Return state if address points to an import thunk
+ *
+ * An import thunk is setup with a 'jmp' instruction followed by an
+ * absolute address (32bit) or relative offset (64bit) pointing into
+ * the import address table (iat), which is partially maintained by
+ * the runtime linker.
+ */
+static BOOL is_import_thunk( void *addr )
+{
+    return *(short *)addr == 0x25ff ? TRUE : FALSE;
+}
+
+/*
+ * Return adress from the import address table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static void *get_address_from_import_address_table( void *iat, DWORD iat_size, void *addr )

```

Comment:
And to be more precise, this function does not return address of function from IAT, but rather from import thunk. So whole check for IAT was there wrong and also comment.

Reply:
32- or 64 bit ?

Reply:
Both, see previous AppVeyor results.

Reply:
32bit: https://ci.appveyor.com/project/dlfcn-win32/dlfcn-win32/builds/36819915/job/dh5ows33l7gtyiq8
64bit: https://ci.appveyor.com/project/dlfcn-win32/dlfcn-win32/builds/36819915/job/98itb4i5by5hpixx

Reply:
>      *   4018c8:	ff 25 4a 8a 00 00    	jmpq   *0x8a4a(%rip)        # 40a318 <__imp_dlopen>
This code is the thunk entry.

> *0x8a4a(%rip)        # 40a318 <__imp__dlopen>

The address  0x40a318 points to the symbol `__imp_dlopen`, which is located inside the iat. See this real work example:
```
objdump -D lib/libdll.dll.a
...
Disassembly of section .text:

00000000 <_dlopen>:    --> thunk entry
   0:	ff 25 00 00 00 00    	jmp    *0x0
   6:	90                   	nop
   7:	90                   	nop
...
Disassembly of section .idata$5:

00000000 <__imp__dlopen>:    --> iat entry (see below for more information)
   0:	00 00                	add    %al,(%eax)
	...
```
Compiling test-dladdr.exe with `-Wl,-verbose` shows the used linker script, where the following block is included:
```
.idata BLOCK(__section_alignment__) :
  {
    /* This cannot currently be handled with grouped sections.
        See pe.em:sort_sections.  */
    KEEP (SORT(*)(.idata$2))
    KEEP (SORT(*)(.idata$3))
    /* These zeroes mark the end of the import list.  */
    LONG (0); LONG (0); LONG (0); LONG (0); LONG (0);
    KEEP (SORT(*)(.idata$4))
    __IAT_start__ = .;
    KEEP (SORT(*)(.idata$5))
    __IAT_end__ = .;
    KEEP (SORT(*)(.idata$6))
    KEEP (SORT(*)(.idata$7))
  }
```
This block sets up the iat and specifies that all code defined in section `idata$5` is located inside the iat (between __IAT_start__ and      __IAT_end__) and the imported functions are added to that section as shown below:

objdump -D bin/test-dladdr.exe 
...
0040712c <__IAT_start__>:
  40712c:	08 72 00             	or     %dh,0x0(%edx)
	...

00407134 <__imp__dlopen>:
  407134:	20 72 00             	and    %dh,0x0(%edx)
	...

get_address_from_import_address_table() first  performs the required calculation to get the address of __imp__dlopen and then gets the content  from the iat  with `return *(void **)ptr;`, so I think that the function name and the comment is correct  or do I misunderstand something ?

Reply:
This did not work with mscv caused by a different binary file layout.